---
title: "Gene Set Mixture Model"
author: "Nicholas Knoblauch"
date: "June 12, 2015"
output: html_document
---


###Introduction

  GWAS, in the most simple context, seeks to identify loci significantly associated with a trait of interest.  In the majority of GWAS, variants are weighted equally, that is to say, the prior probability that a particular variant contributes to the trait of interest is uniform.  Given that the number of loci in the human genome that are known to very is in the millions, while the number of loci that contribute to a given trait is much smaller, there is an extremely high threshold for significance.
  
This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r}
summary(cars)
```

You can also embed plots, for example:

```{r, echo=FALSE}
plot(cars)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
